Create `crdb_internal.active_server_range_feeds` server side rangefeed internal table

xuchef opened the pull request:
### `sql: introduce crdb_internal.active_server_range_feeds virtual table`

The crdb_internal.active_server_range_feeds virtual table exposes the state of
all active rangefeed registrations on the kv server.

```
> SELECT * FROM crdb_internal.active_server_range_feeds;
  node_id | store_id | range_id | stream_id |     consumer_id     | tags |           catchup_ts           | diff |            created            |            range_start            |             range_end             |            resolved            |  resolved_age   | catchup
----------+----------+----------+-----------+---------------------+------+--------------------------------+------+-------------------------------+-----------------------------------+-----------------------------------+--------------------------------+-----------------+----------
        3 |        3 |       84 |         4 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.42146+00  | /Table/106/1/2515465100960393392  | /Table/106/1/4192441834933988992  | 1755887782937954994.0000000000 | 00:00:03.854878 |    f
        3 |        3 |       82 |         5 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.43205+00  | /Table/106/1/4192441834933988992  | /Table/106/1/5869418568907584592  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       82 |         3 | 1100406207275204614 |      | 1755887723981883049.0000000000 |  f   | 2025-08-22 18:35:24.081502+00 | /Table/106/1/4192441834933988992  | /Table/106/1/5869418568907584592  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       82 |         4 | 1100406209894219782 |      | 1755887724782129345.0000000000 |  f   | 2025-08-22 18:35:24.954521+00 | /Table/106/1/4192441834933988992  | /Table/106/1/5869418568907584592  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       77 |         6 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.44254+00  | /Table/106/1/5869418568907584592  | /Table/106/1/7546395302881180192  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       77 |         5 | 1100406209894219782 |      | 1755887724782129345.0000000000 |  f   | 2025-08-22 18:35:24.964589+00 | /Table/106/1/5869418568907584592  | /Table/106/1/7546395302881180192  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       96 |         7 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.452605+00 | /Table/106/1/7546395302881180192  | /Table/106/2                      | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
(7 rows)
```

Resolves: https://github.com/cockroachdb/cockroach/issues/146462
Epic: CRDB-50533
Release note: None

### `kvcoord,sql: store rangefeed consumerID and streamID on kv client`

There is a 1:1 correspondence between a `PartialRangefeed` on the kv client and
a rangefeed `registration` on the kv server. The state of these in-memory
structs are exposed as rows in the crdb_internal.active_range_feeds and
crdb_internal.active_server_range_feeds virtual tables, respectively.

However, there was no simple way of mapping a row in one table to the other.

To address this, this patch stores the rangefeed consumer_id and stream_id on
the kv client, and adds them as columns to crdb_internal.active_range_feeds.

Now, the columns (consumer_id, stream_id) provide a 1:1 mapping between rows in
crdb_internal.active_range_feeds and crdb_internal.active_server_range_feeds.

Part of: https://github.com/cockroachdb/cockroach/issues/146462
Epic: CRDB-50533
Release note: None

### `server,kvcoord: set default rangefeed consumerID on kv client`

Most rangefeeds are created to service a changefeed and have their consumerID
set to the changfeed's JobID. However, system rangefeeds (e.g.
system-config-cache) are not part of a changefeed job and thus have an
uninitialized consumerID.

Previously, when the kv client started a rangefeed, we allowed for the
consumerID to be left uninitialized. The kv server, upon receiving a rangefeed
request with an uninitialized consumerID, would generate a random one and store
it locally, never relaying it back to the client.

We introduced the set of keys (consumer_id, stream_id) as a mapping between rows
in the crdb_internal.active_range_feeds and
crdb_internal.active_server_range_feeds virtual tables. However, this mapping
doesn't work for system rangefeeds since their consumerID field is left
uninitialized on the kv client.

To address this, this patch shifts the default consumerID generation logic to
the kv client, ensuring that every rangefeed request has consumerID initialized.

Now, all rangefeeds can be uniquely identifed on the kv client and server with
the keys (consumer_id, stream_id).

Part of: https://github.com/cockroachdb/cockroach/issues/146462
Epic: CRDB-50533
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147076)
<!-- Reviewable:end -->

xuchef commented:
Hey @stevendanna (pinging you cause Wenyi is out the rest of the week) I was hoping to get some guidance on where we're trying to go with this issue.

Here's what we've got so far in terms of functionality (code is still very rough).

On the `/inspectz/rangefeed` route we have an output that looks like:
```json
{
  "rangefeed_info_per_store": [
    {
      "node_id": 1,
      "store_id": 1,
      "rangefeed": [
        {
          "span": {
            "key": "jg==",
            "end_key": "jw=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079469192e+18
          }
        },
        {
          "span": {
            "key": "jg==",
            "end_key": "jw=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080924056e+18
          }
        },
        {
          "span": {
            "key": "jg==",
            "end_key": "jw=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082066583e+18
          }
        },
        {
          "span": {
            "key": "wg==",
            "end_key": "ww=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080650508e+18
          }
        },
        {
          "span": {
            "key": "wg==",
            "end_key": "ww=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080997759e+18
          }
        },
        {
          "span": {
            "key": "wg==",
            "end_key": "ww=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082126538e+18
          }
        },
        {
          "span": {
            "key": "kIk=",
            "end_key": "kIo="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.74784308065083e+18
          }
        },
        {
          "span": {
            "key": "kIk=",
            "end_key": "kIo="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080998698e+18
          }
        },
        {
          "span": {
            "key": "kIk=",
            "end_key": "kIo="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082127539e+18
          }
        },
        {
          "span": {
            "key": "ug==",
            "end_key": "uw=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157509233e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080650711e+18
          }
        },
        {
          "span": {
            "key": "ug==",
            "end_key": "uw=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157509233e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080998452e+18
          }
        },
        {
          "span": {
            "key": "ug==",
            "end_key": "uw=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157509233e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082127267e+18
          }
        },
        {
          "span": {
            "key": "t4k=",
            "end_key": "t4o="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079469123e+18
          }
        },
        {
          "span": {
            "key": "t4k=",
            "end_key": "t4o="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080923984e+18
          }
        },
        {
          "span": {
            "key": "t4k=",
            "end_key": "t4o="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082066555e+18
          }
        },
        {
          "span": {
            "key": "iw==",
            "end_key": "jA=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {

          }
        },
        {
          "span": {
            "key": "iw==",
            "end_key": "jA=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079599522e+18
          }
        },
        {
          "span": {
            "key": "iw==",
            "end_key": "jA=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079870704e+18
          }
        },
        {
          "span": {
            "key": "iw==",
            "end_key": "jA=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {

          }
        },
        {
          "span": {
            "key": "iw==",
            "end_key": "jA=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080943444e+18
          }
        },
        {
          "span": {
            "key": "iw==",
            "end_key": "jA=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {

          }
        },
        {
          "span": {
            "key": "iw==",
            "end_key": "jA=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082084018e+18
          }
        },
        {
          "span": {
            "key": "too=",
            "end_key": "tos="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079599069e+18
          }
        },
        {
          "span": {
            "key": "too=",
            "end_key": "tos="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080943056e+18
          }
        },
        {
          "span": {
            "key": "too=",
            "end_key": "tos="
          },
          "resolved_ts": {
            "wall_time": 1.747843157323688e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082083647e+18
          }
        },
        {
          "span": {
            "key": "jQ==",
            "end_key": "jg=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079599522e+18
          }
        },
        {
          "span": {
            "key": "jQ==",
            "end_key": "jg=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079870704e+18
          }
        },
        {
          "span": {
            "key": "jQ==",
            "end_key": "jg=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080943444e+18
          }
        },
        {
          "span": {
            "key": "jQ==",
            "end_key": "jg=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082084018e+18
          }
        },
        {
          "span": {
            "key": "qA==",
            "end_key": "qQ=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843079870704e+18
          }
        },
        {
          "span": {
            "key": "nA==",
            "end_key": "nQ=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080622592e+18
          }
        },
        {
          "span": {
            "key": "nA==",
            "end_key": "nQ=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843080985356e+18
          }
        },
        {
          "span": {
            "key": "nA==",
            "end_key": "nQ=="
          },
          "resolved_ts": {
            "wall_time": 1.747843157250458e+18
          },
          "catchup_ts": {
            "wall_time": 1.747843082113897e+18
          }
        }
      ]
    }
  ]
}
```

In the sql shell we have
```sql
select * from crdb_internal.active_server_range_feeds;
```

```
  id | store_id | consumer_id | stream_id | span_start  |  span_end   |          resolved_ts           |           catchup_ts
-----+----------+-------------+-----------+-------------+-------------+--------------------------------+---------------------------------
   1 |        1 |           0 |         0 | /Table/6    | /Table/7    | 1747843340411888000.0000000000 | 1747843079469192000.0000000000
   1 |        1 |           0 |         0 | /Table/6    | /Table/7    | 1747843340411888000.0000000000 | 1747843080924056000.0000000000
   1 |        1 |           0 |         0 | /Table/6    | /Table/7    | 1747843340411888000.0000000000 | 1747843082066583000.0000000000
   1 |        1 |           0 |         0 | /Table/58   | /Table/59   | 1747843340412468000.0000000000 | 1747843080650508000.0000000000
   1 |        1 |           0 |         0 | /Table/58   | /Table/59   | 1747843340412468000.0000000000 | 1747843080997759000.0000000000
   1 |        1 |           0 |         0 | /Table/58   | /Table/59   | 1747843340412468000.0000000000 | 1747843082126538000.0000000000
   1 |        1 |           0 |         0 | /Table/8/1  | /Table/8/2  | 1747843340411940000.0000000000 | 1747843080650830000.0000000000
   1 |        1 |           0 |         0 | /Table/8/1  | /Table/8/2  | 1747843340411940000.0000000000 | 1747843080998698000.0000000000
   1 |        1 |           0 |         0 | /Table/8/1  | /Table/8/2  | 1747843340411940000.0000000000 | 1747843082127539000.0000000000
   1 |        1 |           0 |         0 | /Table/5    | /Table/6    | 1747843340411888000.0000000000 | 1747843079599522000.0000000000
   1 |        1 |           0 |         0 | /Table/5    | /Table/6    | 1747843340411888000.0000000000 | 1747843079870704000.0000000000
   1 |        1 |           0 |         0 | /Table/5    | /Table/6    | 1747843340411888000.0000000000 | 1747843080943444000.0000000000
   1 |        1 |           0 |         0 | /Table/5    | /Table/6    | 1747843340411888000.0000000000 | 1747843082084018000.0000000000
   1 |        1 |           0 |         0 | /Table/32   | /Table/33   | 1747843340412468000.0000000000 | 1747843079870704000.0000000000
   1 |        1 |           0 |         0 | /Table/20   | /Table/21   | 1747843340411940000.0000000000 | 1747843080622592000.0000000000
   1 |        1 |           0 |         0 | /Table/20   | /Table/21   | 1747843340411940000.0000000000 | 1747843080985356000.0000000000
   1 |        1 |           0 |         0 | /Table/20   | /Table/21   | 1747843340411940000.0000000000 | 1747843082113897000.0000000000
   1 |        1 |           0 |         0 | /Table/50   | /Table/51   | 1747843340412468000.0000000000 | 1747843080650711000.0000000000
   1 |        1 |           0 |         0 | /Table/50   | /Table/51   | 1747843340412468000.0000000000 | 1747843080998452000.0000000000
   1 |        1 |           0 |         0 | /Table/50   | /Table/51   | 1747843340412468000.0000000000 | 1747843082127267000.0000000000
   1 |        1 |           0 |         0 | /Table/47/1 | /Table/47/2 | 1747843340412468000.0000000000 | 1747843079469123000.0000000000
   1 |        1 |           0 |         0 | /Table/47/1 | /Table/47/2 | 1747843340412468000.0000000000 | 1747843080923984000.0000000000
   1 |        1 |           0 |         0 | /Table/47/1 | /Table/47/2 | 1747843340412468000.0000000000 | 1747843082066555000.0000000000
   1 |        1 |           0 |         0 | /Table/3    | /Table/4    | 1747843340412468000.0000000000 |                   0.0000000000
   1 |        1 |           0 |         0 | /Table/3    | /Table/4    | 1747843340412468000.0000000000 | 1747843079599522000.0000000000
   1 |        1 |           0 |         0 | /Table/3    | /Table/4    | 1747843340412468000.0000000000 | 1747843079870704000.0000000000
   1 |        1 |           0 |         0 | /Table/3    | /Table/4    | 1747843340412468000.0000000000 |                   0.0000000000
   1 |        1 |           0 |         0 | /Table/3    | /Table/4    | 1747843340412468000.0000000000 | 1747843080943444000.0000000000
   1 |        1 |           0 |         0 | /Table/3    | /Table/4    | 1747843340412468000.0000000000 |                   0.0000000000
   1 |        1 |           0 |         0 | /Table/3    | /Table/4    | 1747843340412468000.0000000000 | 1747843082084018000.0000000000
   1 |        1 |           0 |         0 | /Table/46/2 | /Table/46/3 | 1747843340412468000.0000000000 | 1747843079599069000.0000000000
   1 |        1 |           0 |         0 | /Table/46/2 | /Table/46/3 | 1747843340412468000.0000000000 | 1747843080943056000.0000000000
   1 |        1 |           0 |         0 | /Table/46/2 | /Table/46/3 | 1747843340412468000.0000000000 | 1747843082083647000.0000000000

```

I've been using a test cluster created using roachprod to get those results.

For reference, Wenyi's dropped some things on my [welcome doc](https://docs.google.com/document/d/1dOMQwRZ5EwzV-0ob3EGgoxzbAdGIIfWj0uBAJ1UZnlk/edit?usp=sharing). However, I'm having a bit of trouble making sense of everything and not sure what the scope of this task is.

Would appreciate it if you could also share some context on how we use these internal tables so we can figure out relative priorities.

Thanks in advance!

stevendanna commented:
@xuchef My apologies for the delay here.

> Would appreciate it if you could also share some context on how we use these internal tables so we can figure out relative priorities.

We use `crdb_internal` virtual tables to expose runtime state about various subsystems. We often use them during debugging or development. Our _most_ common debugging task related to rangefeeds is trying to answer _why_ a downstream rangefeed user is reporting that their rangefeed is "stuck" or "slow".

What this typically means is that some consumer of a rangefeed such as changefeeds (or a rangefeedcache) is not seeing its frontier advance.  The frontier being the resolved timestamp across all of the rangefeeds consumed by that changefeed.  The task then is to determine which part of that frontier isn't advancing and why.  95% of the time the cause is a slow consumer.  But, 5% of the time it is a server-side issue or bug. The goal of _all_ of our observability work in the near term is to more quickly be able to decide which of those buckets we are in and start investigating.  For example, the log line you added in your first contribution will help determine if we are in the 95% bucket (as a follow-up there, it might be even more helpful to have that as a _metric_).  

This table will be another tool. It will serve as the source data for queries that give us information to narrow down where a problem might be. We might use to construct queries to determine if rangefeeds on a particular node all happen to be behind, if the only rangefeeds that are behind are for a particular consumer, or in the very common case to show that there is no server-side problem at all.

Now, at this point you may be feeling that this was a lot of words to read and not a lot of guidance. And you are right there. Frankly, we don't know what we want in this table. But that is OK, we almost never know what we want in the first version of these tables. Luckily, crdb_internal virtual tables are not part of our compatibility guarantee. We can change them as we learn about what is useful.  Right now we should focus on 

1) Making sure that the table would be easy to extend as we learn more about what we need. 
2) Making sure that the table is always _safe_ to query.

I think there are really two options for where this information would be stored, the first is at the level of the mux rangefeed call and the second is at the level of the registration which you've done here.  By placing it near the registration I _think_ makes sense because then our "blind spot" will be the "sender" (the code between the registration and the gRPC client that connected to us). But, we'll have some insight into whether the problem might be there because we will then have information about both the input to the sender (the registration) and the output (the dist sender client). 

I haven't yet looked closely at the implementation to analyze (2) yet.

Commit:
kvserver/rangefeed,inspectz: add inspectz support for rangefeeds

This patch introduces the /inspectz/rangefeeds endpoint, which exposes the
state of all actively running rangefeed registrations on the kv server.

Part of: #146462
Epic: CRDB-50533
Release note: None

Commit:
sql: introduce crdb_internal.active_server_range_feeds virtual table

The crdb_internal.active_server_range_feeds virtual table exposes the state of
all active rangefeed registrations on the kv server.

```
> SELECT * FROM crdb_internal.active_server_range_feeds;
  node_id | store_id | range_id | stream_id |     consumer_id     | tags |           catchup_ts           | diff |            created            |            range_start            |             range_end             |            resolved            |  resolved_age   | catchup
----------+----------+----------+-----------+---------------------+------+--------------------------------+------+-------------------------------+-----------------------------------+-----------------------------------+--------------------------------+-----------------+----------
        3 |        3 |       84 |         4 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.42146+00  | /Table/106/1/2515465100960393392  | /Table/106/1/4192441834933988992  | 1755887782937954994.0000000000 | 00:00:03.854878 |    f
        3 |        3 |       82 |         5 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.43205+00  | /Table/106/1/4192441834933988992  | /Table/106/1/5869418568907584592  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       82 |         3 | 1100406207275204614 |      | 1755887723981883049.0000000000 |  f   | 2025-08-22 18:35:24.081502+00 | /Table/106/1/4192441834933988992  | /Table/106/1/5869418568907584592  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       82 |         4 | 1100406209894219782 |      | 1755887724782129345.0000000000 |  f   | 2025-08-22 18:35:24.954521+00 | /Table/106/1/4192441834933988992  | /Table/106/1/5869418568907584592  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       77 |         6 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.44254+00  | /Table/106/1/5869418568907584592  | /Table/106/1/7546395302881180192  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       77 |         5 | 1100406209894219782 |      | 1755887724782129345.0000000000 |  f   | 2025-08-22 18:35:24.964589+00 | /Table/106/1/5869418568907584592  | /Table/106/1/7546395302881180192  | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
        3 |        3 |       96 |         7 | 1100406204630007814 |      | 1755887723174292752.0000000000 |  f   | 2025-08-22 18:35:23.452605+00 | /Table/106/1/7546395302881180192  | /Table/106/2                      | 1755887782817691099.0000000000 | 00:00:03.975142 |    f
(7 rows)
```

Resolves: #146462
Epic: CRDB-50533
Release note: None

Commit:
kvcoord,sql: store rangefeed consumerID and streamID on kv client

There is a 1:1 correspondence between a `PartialRangefeed` on the kv client and
a rangefeed `registration` on the kv server. The state of these in-memory
structs are exposed as rows in the crdb_internal.active_range_feeds and
crdb_internal.active_server_range_feeds virtual tables, respectively.

However, there was no simple way of mapping a row in one table to the other.

To address this, this patch stores the rangefeed consumer_id and stream_id on
the kv client, and adds them as columns to crdb_internal.active_range_feeds.

Now, the columns (consumer_id, stream_id) provide a 1:1 mapping between rows in
crdb_internal.active_range_feeds and crdb_internal.active_server_range_feeds.

Part of: #146462
Epic: CRDB-50533
Release note: None

xuchef commented on pkg/kv/kvserver/rangefeed/registry.go:
```diff
@@ -96,6 +116,67 @@ type baseRegistration struct {
+func (r *baseRegistration) getRangefeedState() rangefeedpb.RangefeedState {
+	buf := logtags.FromContext(r.streamCtx)
+
+	// TODO(mxu): the alternative would be setting these values at
+	// registration-time, which would mean they won't have to be re-fetched for
+	// every Inspectz query, but it would also mean that it's unneccessarily
+	// fetched, as we only use these values for obs (via this function).
+	getInt64Tag := func(key string) int64 {
+		if buf == nil {
+			return 0
+		}
+		t, ok := buf.GetTag(key)
+		if !ok {
+			return 0
+		}
+		v, ok := t.Value().(int64)
+		if !ok {
+			return 0
+		}
+		return v
+	}
+
+	var tags string
+	if buf != nil {
+		tags = buf.String()
+	}
+
+	return rangefeedpb.RangefeedState{
+		ConsumerID:     getInt64Tag("cid"),
+		StreamID:       getInt64Tag("sid"),
+		Span:           r.span,
+		ResolvedTS:     r.resolvedTimestamp,
+		CatchUpTS:      r.catchUpTimestamp,
+		CatchUpRunning: r.catchUpRunning,
+		Tags:           tags,
+		LastEventTS:    &r.lastValueSent,
+		RangeID:        r.rangeID,
+		Diff:           r.withDiff,
+		Created:        &r.createdTime,
+	}
```
grabbing these from the tags feels a bit dirty, but the alternative was to drill them down as params which would require changing signatures. is the performance cost of GetTag from a context real? if so, is it better to subject every registration to it unnecessarily (do it at registration-time) or don't cache it and do it on each request (as it is now)?

we could also just drop these two columns and just have the consumerID and streamID as text values in tags (and then we wouldn't need 2c7a01d4c0c0eb7115757a298e99e51ab9f827bd either).

xuchef replied:
Cool, we're now passing streamID and consumerID as arguments. Had to update a few tests to use dummy values.

I considered obtaining these ID's from `stream`, but it's type is the `Stream` interface, and thus isn't currently tied to the `PerRangeEventSink` struct that stores these fields (we make use of this fact in testing where we have mock stream implementations). If we did want to get the ID's from `stream`, one way would be to make a new "Metadata" interface that exposes a getter, but that would still require some ugly typecasting at the call site (`if s, ok := stream.(Metadata)`). Also, this would mean adding `consumerID` to `PerRangeEventSink` even though we only annotate events with `streamID` and `rangeID`.

Anyways, TLDR: it's a little redundant having `streamID` and `consumerID` as separate args because they really belong to the `stream` but it avoids finicking with types, so I went with the simplest solution for now.

Happy to discuss other options though if this isn't what we're looking for.
<!-- Reviewable comment -OVELT9ZCGzZHW_QHZCB:bnzvisv -->
<!-- Sent from Reviewable.io -->

Commit:
server,kvcoord: set default rangefeed consumerID on kv client

Most rangefeeds are created to service a changefeed and have their consumerID
set to the changfeed's JobID. However, system rangefeeds (e.g.
system-config-cache) are not part of a changefeed job and thus have an
uninitialized consumerID.

Previously, when the kv client started a rangefeed, we allowed for the
consumerID to be left uninitialized. The kv server, upon receiving a rangefeed
request with an uninitialized consumerID, would generate a random one and store
it locally, never relaying it back to the client.

We introduced the set of keys (consumer_id, stream_id) as a mapping between rows
in the crdb_internal.active_range_feeds and
crdb_internal.active_server_range_feeds virtual tables. However, this mapping
doesn't work for system rangefeeds since their consumerID field is left
uninitialized on the kv client.

To address this, this patch shifts the default consumerID generation logic to
the kv client, ensuring that every rangefeed request has consumerID initialized.

Now, all rangefeeds can be uniquely identifed on the kv client and server with
the keys (consumer_id, stream_id).

Part of: #146462
Epic: CRDB-50533
Release note: None

mw5h commented:
SQL changes appear to be in Foundations' neck of the woods, so I'll defer to them.

stevendanna reviewed (commented):
Reviewable has gotten incredibly confused about my review state so I'm going to flush my current comments so I don't lose them.

> Reviewed 11 of 20 files at r1, 14 of 14 files at r4, 1 of 18 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-:-OV7NaYxA7TDifO39HEn:b-8frk37)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @mw5h, @wenyihu6, and @xuchef)

___
*[`pkg/kv/kvserver/store.go` line 4248 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7Gjrs1z7pUluCk1WV:-OV7Gjrs1z7pUluCk1WW:b98mgq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/store.go#L4248)):*
> ```Go
> 
> func (s *Store) VisitRangefeeds() (res rangefeedpb.RangefeedInfoPerStore) {
> 	var rangeIDs []roachpb.RangeID // avoid alloc on each call
> ```

I'm a little confused by this comment. The function below is only called once.

It might also be good to have a comment here pointing out that we aren't using VisitReplicas here because we only want replicas with rangefeeds on them.
___
*[`pkg/kv/kvserver/store.go` line 4259 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7IwR3DKQIbPUwju6W:-OV7IwR3DKQIbPUwju6X:boceh4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/store.go#L4259)):*
> ```Go
> 	}
> 
> 	res.StoreID = *s.Ident
> ```

Is this right? I thought Ident was a struct with ClusterID, StoreID, and NodeID.  It seems like either this assignment is wrong or res.StoreID should be renamed.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 84 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7MVNi8dI-BeB-1D_t:-OV7MVNi8dI-BeB-1D_u:b-pxsx6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/registry.go#L84)):*
> ```Go
> 	getCatchUpRunning() bool
> 	// getLastValueSent returns the lastValueSent field of the registration.
> 	getLastValueSent() time.Time
> ```

Do we definitely need these accessors?  I don't see most of them being called.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7Jpvg4rZ51LPN42zW:-OV7Jpvg4rZ51LPN42zX:b-3xmbz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/registry.go#L146)):*
> ```Go
> 	// registration-time, which would mean they won't have to be re-fetched for
> 	// every Inspectz query, but it would also mean that it's unneccessarily
> 	// fetched, as we only use these values for obs (via this function).
> ```

If we keep these, we could perhaps keep a *string somewhere such that we only have to pull them from the context the first time Inspectz is called.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 500 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7LY-oBZeiMEUfMJXk:-OV7LY-oBZeiMEUfMJXl:b-yxe0g3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/registry.go#L500)):*
> ```Go
> 	ctx context.Context, span roachpb.Span,
> ) (states []rangefeedpb.RangefeedState) {
> 	reg.forOverlappingRegs(ctx, span, func(r registration) (bool, *kvpb.Error) {
> ```

Can we pre-allocate `states` using reg.tree.Len()?
___
*[`pkg/kv/kvserver/rangefeed/unbuffered_registration.go` line 365 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7M9Fh-dVCWurMtcD6:-OV7M9Fh-dVCWurMtcD7:b-1jltzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/unbuffered_registration.go#L365)):*
> ```Go
> 	}
> 	start := timeutil.Now()
> 	ubr.baseRegistration.catchUpRunning = true
> ```

Is this going to trip the race detector when you try to read from it?


<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/sql/crdb_internal.go:
```diff
@@ -7314,6 +7316,92 @@ CREATE TABLE crdb_internal.active_range_feeds (
+	comment: `node-level table listing all currently running rangefeeds on the kv server`,
+	schema: `
+CREATE TABLE crdb_internal.active_server_range_feeds (
+  consumer_id INT,
+  stream_id INT,
```
Let's re-arrange these fields a bit. I would expect `node_id, store_id, range_id, stream_id` to be first.

xuchef replied:
Fixed. Also reordered the fields in the client table.
<!-- Reviewable comment -OV91A2qErVFPYem07cW:b-virbdn -->
<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/sql/crdb_internal.go:
```diff
@@ -7314,6 +7316,92 @@ CREATE TABLE crdb_internal.active_range_feeds (
+  store_id INT,
+  range_id INT,
+  created TIMESTAMPTZ,
+  range_start STRING,
+  range_end STRING,
```
Is this actually the start of the range or is it the start key of the span for the registration? If the latter, I would name these start_key and end_key, I would have then be the raw key bytes.

xuchef replied:
These are the keys of the span for the registration.

From our discussion in the meeting, `start_key` and `end_key` would be better names, but to keep it consistent with the client side table, we'll keep the misnomers `range_start` and `range_end` for now. Added a TODO for ssd to change later.
<!-- Reviewable comment -OV9-J6L8zEZBW_zJOP9:b-xokqew -->
<!-- Sent from Reviewable.io -->

stevendanna commented on pkg/server/node.go:
```diff
@@ -2253,36 +2245,31 @@ func (n *Node) muxRangeFeed(muxStream kvpb.RPCInternal_MuxRangeFeedStream) error
+			//   - For rangefeeds created to service a changefeed, the ConsumerID is
+			//     the JobID of the changefeed.
+			//   - For system rangefeeds, the ConsumerID is a random, unique ID.
+			if consumerID == 0 {
+				consumerID = req.ConsumerID
+			} else if consumerID != req.ConsumerID {
```
I think to handle mixed-version clusters, we'll still need to handle the case of req.ConsumerID being unset.

xuchef replied:
Gotcha. We've gotta keep the default id generation code on the server to support rolling upgrades where the client is on an older version. Undid the deletion, and added a comment explaining our reasoning.

Made sure to set `req.ConsumerID` to be `consumerID` in the case that they differ, as that's the value we use when we create registrations.
<!-- Reviewable comment -OV8xpMk77M2QmT0K876:bxde0wg -->
<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
Thanks for working on this! Overall, this has the shape I expected.  I've left a handful of comments that I think we need to sort out, but none are too major.

xuchef reviewed (commented):
Thanks for the comments, pushed changes addressing them. Let me know if anything else stands out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-:-OVF1MlF81kyQL5Kqe5W:b-xr4r97)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @mw5h, @stevendanna, and @wenyihu6)

___
*[`pkg/kv/kvserver/store.go` line 4248 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7Gjrs1z7pUluCk1WV:-OV8pH1s6PtED37Vv_UL:bw0we1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/store.go#L4248)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'm a little confused by this comment. The function below is only called once.

It might also be good to have a comment here pointing out that we aren't using VisitReplicas here because we only want replicas with rangefeeds on them.
</blockquote></details>

Yeah what I had before didn't make sense. Updated the code and comment.
___
*[`pkg/kv/kvserver/store.go` line 4259 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7IwR3DKQIbPUwju6W:-OV8PV82CPPV31IGVc89:b-xafshb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/store.go#L4259)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Is this right? I thought Ident was a struct with ClusterID, StoreID, and NodeID.  It seems like either this assignment is wrong or res.StoreID should be renamed.
</blockquote></details>

It should be renamed. I was using storeliveness as an example where `StoreIdent` is [called](https://github.com/cockroachdb/cockroach/blob/a5beea5bb2215f9851e557806d16265885670fd0/pkg/kv/kvserver/storeliveness/storelivenesspb/service.proto#L171) `StoreID` but it also has a [field](https://github.com/cockroachdb/cockroach/blob/a5beea5bb2215f9851e557806d16265885670fd0/pkg/kv/kvserver/storeliveness/storelivenesspb/service.proto#L13-L20) `StoreID`. I also found it quite confusing.

Is `res.StoreIdent` a better name?
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 84 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7MVNi8dI-BeB-1D_t:-OV8nOtr81qNS3bhAgJS:bcc1n3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/registry.go#L84)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Do we definitely need these accessors?  I don't see most of them being called.
</blockquote></details>

Nope, they're not used at all. Removed.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7Jpvg4rZ51LPN42zW:-OVDflut-g1ZS5W8SD9_:btocybr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/registry.go#L146)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

If we keep these, we could perhaps keep a *string somewhere such that we only have to pull them from the context the first time Inspectz is called.
</blockquote></details>

Not parsing from the tags anymore. Resolved.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 500 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7LY-oBZeiMEUfMJXk:-OV8rfsCCDWWUycZgnNK:b11g8nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/registry.go#L500)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Can we pre-allocate `states` using reg.tree.Len()?
</blockquote></details>

Currently, the only place we call this function is`CollectAllStates` in which case the pre-allocation makes perfect sense. In other future cases, we'd be pre-allocating the upper bound - is that ok? We could delete this intermediate function for now.
___
*[`pkg/kv/kvserver/rangefeed/unbuffered_registration.go` line 365 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OV7M9Fh-dVCWurMtcD6:-OVEai-VDB1mW2oXbu1j:b-v9aqv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/rangefeed/unbuffered_registration.go#L365)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Is this going to trip the race detector when you try to read from it?
</blockquote></details>

Ah yes, it does indeed, confirmed with a test. Same thing for `resolvedTimestamp` and `lastValueSent`. Refactored to move these mutable fields under the mutex of `bufferedRegistration` and `unbufferedRegistration`. Now, `baseRegistration` only contains the immutable fields set at init-time.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147076)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @mw5h, and @stevendanna)

___
*[`pkg/kv/kvserver/rangefeed/buffered_registration.go` line 350 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJmSF88YaO-8DYkLao:-OVJmSF88YaO-8DYkLap:bgt4ux) ([raw file](https://github.com/cockroachdb/cockroach/blob/be22ad9acf7f36d10fe63cef057f906d384f1ec1/pkg/kv/kvserver/rangefeed/buffered_registration.go#L350)):*
> ```Go
> 	br.mu.Lock()
> 	br.mu.catchUpRunning = true
> 	br.mu.Unlock()
> ```

nit: mind moving these to helper functions with a defer on Unlock as well 
___
*[`pkg/kv/kvserver/rangefeed/rangefeedpb/service.proto` line 17 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJnLX795C9CFzSK_lr:-OVJnLX795C9CFzSK_ls:b-tc61d) ([raw file](https://github.com/cockroachdb/cockroach/blob/be22ad9acf7f36d10fe63cef057f906d384f1ec1/pkg/kv/kvserver/rangefeed/rangefeedpb/service.proto#L17)):*
> ```proto
> // RangefeedState represents the state of a single active rangefeed
> // registration.
> message RangefeedState {
> ```

Would you mind moving the proto changes to a separate PR? We can merge them in once we agree first. 
___
*[`pkg/kv/kvserver/store.go` line 4251 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJhmSUB4oHXhQDveJ4:-OVJhmSUB4oHXhQDveJ5:b-19d2jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/store.go#L4251)):*
> ```Go
> 	updateRangeIDs := func() []roachpb.RangeID {
> 		rangeIDs = rangeIDs[:0]
> 		s.rangefeedReplicas.Lock()
> ```

nit: can we use derfer for Unlock()
___
*[`pkg/kv/kvserver/rangefeed/unbuffered_registration.go` line 182 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJiyqr1aisSPIFhGKT:-OVJiyqr1aisSPIFhGKU:b-2fyu77) ([raw file](https://github.com/cockroachdb/cockroach/blob/be22ad9acf7f36d10fe63cef057f906d384f1ec1/pkg/kv/kvserver/rangefeed/unbuffered_registration.go#L182)):*
> ```Go
> 
> 	if event.Checkpoint != nil {
> 		ubr.mu.resolvedTimestamp = event.Checkpoint.ResolvedTS
> ```

nit: we might want to revisit this function since it's on a hot path, but I'm not clear on the performance cost is of maintaining state like this on every registration. 


<!-- Sent from Reviewable.io -->

xuchef reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147076)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, @mw5h, @stevendanna, and @wenyihu6)

___
*[`pkg/kv/kvserver/store.go` line 4251 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJhmSUB4oHXhQDveJ4:-OY7ldhR0P2XCGT_WTUJ:b1gayl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7765b499259786c2fec6c7780f57e7bcfe354d68/pkg/kv/kvserver/store.go#L4251)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: can we use derfer for Unlock()
</blockquote></details>

Talked offline. Makes more sense to do it like this because we don't wanna hold onto the lock for longer than we need.
___
*[`pkg/kv/kvserver/rangefeed/buffered_registration.go` line 350 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJmSF88YaO-8DYkLao:-OY7lLrwEEJU3rzNetrf:bnytqui) ([raw file](https://github.com/cockroachdb/cockroach/blob/be22ad9acf7f36d10fe63cef057f906d384f1ec1/pkg/kv/kvserver/rangefeed/buffered_registration.go#L350)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: mind moving these to helper functions with a defer on Unlock as well
</blockquote></details>

Done, added a `setCatchupRunning` setter.
___
*[`pkg/kv/kvserver/rangefeed/unbuffered_registration.go` line 182 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJiyqr1aisSPIFhGKT:-OY7leUZE5bg7Nhov4C5:bgkda0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/be22ad9acf7f36d10fe63cef057f906d384f1ec1/pkg/kv/kvserver/rangefeed/unbuffered_registration.go#L182)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: we might want to revisit this function since it's on a hot path, but I'm not clear on the performance cost is of maintaining state like this on every registration.
</blockquote></details>

Ran a quick test putting this section in a function

```diff
diff --git a/pkg/kv/kvserver/rangefeed/unbuffered_registration.go b/pkg/kv/kvserver/rangefeed/unbuffered_registration.go
index 0ae5e517e03..b6044344b4a 100644
--- a/pkg/kv/kvserver/rangefeed/unbuffered_registration.go
+++ b/pkg/kv/kvserver/rangefeed/unbuffered_registration.go
@@ -164,6 +164,15 @@ func (ubr *unbufferedRegistration) getRangefeedState() rangefeedpb.RangefeedStat
        }
 }
 
+func (ubr *unbufferedRegistration) updateStateLocked(event *kvpb.RangeFeedEvent) {
+       if event.Checkpoint != nil {
+               ubr.mu.resolvedTimestamp = event.Checkpoint.ResolvedTS
+       }
+       if event.Val != nil || event.SST != nil {
+               ubr.mu.lastValueSent = timeutil.Now()
+       }
+}
+
 // publish sends a single event to this registration. It is called by the
 // processor if the event overlaps the span this registration is interested in.
 // Events are either stored in catchUpBuf or sent to BufferedStream directly,
@@ -179,13 +188,7 @@ func (ubr *unbufferedRegistration) publish(
        }
 
        ubr.assertEvent(ctx, event)
-
-       if event.Checkpoint != nil {
-               ubr.mu.resolvedTimestamp = event.Checkpoint.ResolvedTS
-       }
-       if event.Val != nil || event.SST != nil {
-               ubr.mu.lastValueSent = timeutil.Now()
-       }
+       ubr.updateStateLocked(event)
 
        strippedEvent := ubr.maybeStripEvent(ctx, event)
```

and running the kv0 workload with a running changefeed

```bash
$ ./bin/roachprod sql $c:6 -- -e 'SET CLUSTER SETTING kv.rangefeed.enabled = true;'
$ roachprod ssh $c:6 -- ./cockroach workload init kv --scatter --splits 500 --drop {pgurl:1-5}
$ roachprod sql $c:6 -- -e "CREATE CHANGEFEED FOR kv.kv INTO 'null://'"
$ roachprod ssh $c:6 -- ./cockroach workload run kv --read-percent 0 --concurrency 320 {pgurl:1-5}
_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
    1.0s        0         7407.9         8059.1     32.5     92.3    130.0    218.1 write
    2.0s        0         9036.3         8545.6     32.5     75.5    100.7    151.0 write
    3.0s        0         8605.3         8565.6     33.6     79.7    104.9    184.5 write
    4.0s        0         8753.9         8626.6     32.5     79.7    104.9    167.8 write
...

```

then captured a 30s cluster-wide CPU profile with

```
https://localhost:29001/debug/pprof/ui/cpu/?node=all&seconds=30&labels=true
```

I don't see much overhead
![Screenshot 2025-08-20 at 2.49.20 PM.png](https://files.reviewableusercontent.io/images/reviewable/61663933/doWDnSDCqn2JGqGaBjJLMK3I/Screenshot%202025-08-20%20at%202.49.20%E2%80%AFPM.png)

So I think we should be good.
___
*[`pkg/kv/kvserver/rangefeed/rangefeedpb/service.proto` line 17 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147076#-OVJnLX795C9CFzSK_lr:-OY7h3be6z6RdhGCv5mH:b-h1ljkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/be22ad9acf7f36d10fe63cef057f906d384f1ec1/pkg/kv/kvserver/rangefeed/rangefeedpb/service.proto#L17)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Would you mind moving the proto changes to a separate PR? We can merge them in once we agree first.
</blockquote></details>

Yep, I addressed your other comments and opened up https://github.com/cockroachdb/cockroach/pull/152187.


<!-- Sent from Reviewable.io -->
